DO NOT MERGE Kymatio frontend/backend -1/2/3D

edouardoyallon opened the pull request:
This PR proposes a new mechanism for computing the scattering, that splits low level and high level calls, for the 2D scattering.

Example of call:

```
from kymatio.scattering2d import Scattering2D_torch as Scattering2D
from kymatio.scattering2d.backend import torch_backend as backend
scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad, backend=backend)
```
or
```
from kymatio.scattering2d import Scattering2D_torch as Scattering2D
scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad)
```

Fix #414 
Fix #398 
Fix #421 
Fix #416
Fix #402 


1. Add a backend and frontend support.
2. Refactors the scattering into a functional manner: frontend, backend and core. Frontend calls core, which relies on backend to compute primitives.
3. Refactors the tests into classes, via pytest. Each submodule of the backend is tested via a pytest class. The frontend is globally tested via a pytest class.
4. Split frontend, backend and core into different subfolders.
5. Remove the previous backend that was using config files, and redefines a backend as a possible argument to the frontend.
6. Support a pure numpy and tensorflow backend.
7. Add a template for the frontend
8. Pass the backend as a namedtuple


__TODO in a future PR__: Updating the documentation, testing the multi GPU support, incorporating more tests for the numpy backend, unifications, more tests for the backend-frontend links(e.g., avoiding a numpy backend on a torch backend), adding the 1D-3D core

eickenberg commented on kymatio/scattering2d/backend/torch_backend.py:
```diff
@@ -171,19 +160,20 @@ def fft(input, direction='C2C', inverse=False):
 
-    if (not input.is_contiguous()):
-        raise (RuntimeError('Tensors must be contiguous!'))
+    if not input.is_contiguous():
```
would be great if `input` was renamed to something that is not a python keyword

eickenberg commented on kymatio/scattering2d/backend/torch_backend.py:
```diff
@@ -255,3 +245,18 @@ def cdgmm(A, B, inplace=False):
 
         return C if not inplace else A.copy_(C)
+
+def new(x, O, M, N):
```
this function still seems super ad hoc. can you write a docstring about what it is for?

edouardoyallon replied:
I removed it. It's simpler now

eickenberg commented on kymatio/scattering2d/backend/torch_backend.py:
```diff
@@ -255,3 +245,18 @@ def cdgmm(A, B, inplace=False):
         return C if not inplace else A.copy_(C)
+
+def new(x, O, M, N):
+    shape = x.shape[:-2] + (O,) + (M,) + (N,)
```
`+ (O, M, N)`

eickenberg commented on kymatio/scattering2d/backend/torch_skcuda_backend.py:
```diff
@@ -345,3 +331,17 @@ def cdgmm(A, B, inplace=False):
         return C
+
+def new(x, O, M, N):
+    shape = x.shape[:-2] + (O,) + (M,) + (N,)
```
again

eickenberg commented on kymatio/scattering2d/filter_bank.py:
```diff
@@ -49,8 +49,6 @@ def filter_bank(M, N, J, L=8):
                 psi_signal_fourier_res = periodize_filter_fft(
                     psi_signal_fourier, res)
-                # add a trailing singleton dimension to mark it as non-complex
-                psi_signal_fourier_res = psi_signal_fourier_res[..., np.newaxis]
```
why is this gone now?

edouardoyallon replied:
it's done here: https://github.com/edouardoyallon/kymatio/blob/kymatio-dev-frontend/kymatio/scattering2d/frontend/torch_frontend.py#L132 this is cleaner! (AFAIR, it was xploiting the fact that the wavelet is real and added a 2nd dimension for torch ; now it just does a potentially complex numpy tensor)

eickenberg replied:
Yeah, I saw that later. Good point

eickenberg commented on kymatio/scattering2d/frontend/__init__.py:
```diff
@@ -0,0 +1,5 @@
+__all__ = ['Scattering2D_torch', 'Scattering2D_numpy']
+
+
+from .torch_frontend import Scattering2D_torch
```
can't really import this if there is no `torch`

eickenberg replied:
or maybe yes, if no torch is imported while running through `torch_frontend`, which I doubt

edouardoyallon replied:
we can, because the frontend import the backend. If no backend is loaded or is specified, it should lead to a bug!

eickenberg commented on kymatio/scattering2d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,115 @@
+from ..utils import compute_padding
+from ..filter_bank import filter_bank
+
+class Scattering2D_numpy(Scattering_numpy):
```
I'd call it `Scattering2DNumpy` or `Scattering2D`

eickenberg commented on kymatio/scattering2d/frontend/torch_frontend.py:
```diff
@@ -0,0 +1,189 @@
+from ...frontend.torch_frontend import Scattering_torch
+
+
+class Scattering2D_torch(Scattering_torch):
```
`Scattering2DTorch` or `Scattering2D`

eickenberg commented on kymatio/scattering2d/frontend/torch_frontend.py:
```diff
@@ -0,0 +1,189 @@
+            if isinstance(c, int):
+                self.phi[c] = torch.from_numpy(self.phi[c]).unsqueeze(-1) # add a trailing singleton dimension to mark
+                # it as non-complex
+                self.register_buffer('tensor' + str(n), self.phi[c])
```
do you get a correspondence between buffer numbers and filters? Otherwise consider using better names ...

edouardoyallon replied:
cf below

eickenberg commented on kymatio/scattering2d/frontend/torch_frontend.py:
```diff
@@ -0,0 +1,189 @@
+                if isinstance(k, int):
+                    self.psi[j][k] = torch.from_numpy(v).unsqueeze(-1) # add a trailing singleton dimension to mark it
+                    # as non-complex
+                    self.register_buffer('tensor' + str(n), self.psi[j][k])
```
better names for the buffers?

edouardoyallon replied:
cf below

eickenberg commented on kymatio/scattering2d/frontend/torch_frontend.py:
```diff
@@ -0,0 +1,189 @@
+        # each time scattering is run, one needs to make sure self.psi and self.phi point to
+        # the correct buffers
+        n = 0
+        buffer_dict = dict(self.named_buffers())
```
ah ok you rely on the exact order. well I guess that works ... but it seems a bit brittle

edouardoyallon replied:
yes, this is a trick. I currently don't have a better idea... buffers are pretty useful.

eickenberg replied:
not advocating against the use of buffers. It is essential that you use them so that `.to` and `.cuda` etc work propertly. I was just thinking that instead of calling them `tensor1` `tensor2` etc, they could be called `filter_j2_l4` or something.

I see that the way you do it works reliably.

Having more explicit namings might help people inspect the buffers better, but it is of course not a must

edouardoyallon replied:
my hope is that we will get rid of this mechanism in a future update of torch... so maybe it could be worth not spending much time on it(also observe it's pretty encapuslated in the frontend)

eickenberg replied:
get rid of buffers? Seems pretty central to pytorch for me.

As mentioned in my previous comments, I don't think a change is necessary - I just wanted to suggest it

edouardoyallon replied:
Sorry - I meant, in a future release of the PyTorch, I hope we will be able to register a list of tensor without having to register each individual tensor..

eickenberg commented on kymatio/scattering2d/tests/test_numpy_scattering2d.py:
```diff
@@ -0,0 +1,56 @@
+from kymatio.scattering2d import Scattering2D_numpy
+import torch
+
+class TestScattering2D_Numpy:
```
remove underscore.

Also why is there are class now? Does it require setup and teardown? We had explicitly removed classes from the tests before

edouardoyallon replied:
no, it was mixed! I think it's more consistent(always the same way to test a routine/class) and cleaner(we can use the same test-template) to do a class per "instance" of the backend

eickenberg replied:
As far as I can see these classes only cause an indent and have no functionality themselves. So they seem completely redundant

edouardoyallon replied:
I slightly disagree: the indent allows structuring the tests and it is pretty readable. Which format would you recommend? How do we delimit tests addressed to different subparts of the code?

eickenberg replied:
I'd recommend adding stuff as

`############## Here we test Scattering2DTorch`

for example.

Unless the classes add something semantically that makes it possible to call e.g. _only the tests from this here class_ using `pytest` command line, I am very against adding such semantic overhead.

edouardoyallon replied:
It is possible: http://doc.pytest.org/en/latest/usage.html

> Another example specifying a test method in the command line:
>
> ```
> pytest test_mod.py::TestClass::test_method
> ```

eickenberg commented on kymatio/scattering2d/tests/test_torch_scattering2d.py:
```diff
@@ -0,0 +1,380 @@
+
+
+# Checked the modulus
+class TestModulus:
```
are the classes necessary?

edouardoyallon replied:
cf above

eickenberg replied:
where? I think we managed to do very well without this overhead

eickenberg commented on kymatio/scattering2d/tests/test_torch_scattering2d.py:
```diff
@@ -0,0 +1,380 @@
+        assert ('must be contiguous' in record.value.args[0])
+
+
+class TestScattering2D_Torch:
```
remove underscore

eickenberg replied:
how is this resolved? This naming scheme hurts any reasonable person's eyes

edouardoyallon replied:
my bad. I'm writing the tf backend, it's solved on my local branch

edouardoyallon commented on kymatio/frontend/torch_frontend.py:
```diff
@@ -0,0 +1,25 @@
+       self.max_order = max_order
+
+   def build(self):
+       self.filters = self.create_and_register_filters()
```
So, the issue is that I'd like a similar code to remain: it's a way to remind the dev that one should call this function in build. I don't know the standard in python, but some reading I had on the internet about oop suggest to use a template-like mechanism.

eickenberg replied:
yeah I get why you did it. just pointing out that it will error.

In general, let's go for a minimal an amount of "oop" stuff as possible

What about the main question, calling `build` from `__init__`?

edouardoyallon replied:
Yes, but as the __init__ might changed, I don't know how to mention the ancestor of the object that its method "__init__" got overloaded. Or should it simply do nothing and still build itself?

eickenberg replied:
well so you are leaving `self.filters = ...` as an instruction to developers. In the same way I'd leave `self.build()` at the end of `__init__`

edouardoyallon replied:
I followed your guideline(please close if u're ok)

eickenberg replied:
Thanks

edouardoyallon commented on kymatio/__init__.py:
```diff
@@ -2,18 +2,22 @@
 __all__ = [
             'Scattering1D',
-            'Scattering2D',
+            'Scattering2D_torch',
```
I think a strength of the current code is that one could use in the same code both torch, tensorflow and numpy even if this is not the standard practice. Also, the config file can be a bit subtle to understand, I think.

eickenberg replied:
I see what you mean, but I am thinking that some easy-to-use defaults might be useful. maybe I am wrong

edouardoyallon replied:
well, `scat=Scattering2DNumpy()` will always work, and if torch is installed, then Scattering2DTorch will too! I think it's more than OKayish, it's more a matter of improving the documentation.

edouardoyallon replied:
Fixed

edouardoyallon commented on kymatio/scattering2d/__init__.py:
```diff
@@ -1,4 +1,5 @@
 
 
-from .scattering2d import Scattering2D
+from .frontend import Scattering2D_torch
```
Well, I slightly disagree: https://github.com/edouardoyallon/kymatio/blob/kymatio-dev-frontend/kymatio/scattering2d/backend/__init__.py#L3-L6 and 
The frontend should require a backend to be loaded, but if no amenable backend is available, then, not only the backend with the missing package is not loaded, but I intended these lines: https://github.com/edouardoyallon/kymatio/blob/kymatio-dev-frontend/kymatio/scattering2d/frontend/torch_frontend.py#L98-L102 to fail.

edouardoyallon replied:
solved now

edouardoyallon commented on kymatio/scattering2d/backend/__init__.py:
```diff
@@ -1,71 +1,16 @@
-    raise ValueError("Backend {} unknown".format(BACKEND))
-
-
+__all__ = ['torch_backend', 'torch_skcuda_backend', 'numpy_backend']
```
Solved. Cf comment above.

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+        self.build()
+
+    def build(self):
+        self.np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))
```
I considered the padding should be defined before calling the object. Is this a wrong thought?

eickenberg replied:
It just seems like a rearrangement of `self.pad_size` or am I getting something wrong?

edouardoyallon replied:
so numpy and torch expects a different padding size parameter, yet the resulting operation should be the same: thus I use an extra variable while inputting the same values as for torch and np. Thus, indeed, it's a simple rearrangement

eickenberg replied:
but this code is unique to the numpy backend as far as I can tell, so you don't have to deal with any dual use cases, so you can set `np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))` in `__call__` directly and use it there

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+        if self.pre_pad:
+            return x
+        else:
+            return np.pad(x, ((0,0), (0,0), self.np_pad[0], self.np_pad[1]), mode='reflect')
```
Currently, yes, as done in torch. I agree the padding should be more flexible later!

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+        -------
+        in_[..., 1:-1, 1:-1]
+    """
+    return in_[..., 1:-1, 1:-1]
```
unpad_ds could be a better naming? (ds standing for downsampled)

eickenberg replied:
If this works for all padding sizes specified above, then I'm fine with it. I know it's like this in the torch code as well. Just a little concerned of discrepancies arising when somebody specifies non-standard padding values

edouardoyallon replied:
The padding is specifically adjusted for the unpadding to work. A bug was actually corrected for the case where the padded value is 2^J.

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if direction == 'C2R':
+        output = np.real(np.fft.ifft2(input, norm=None))*input.shape[-1]*input.shape[-2]
```
would it be OK to do a pure numpy backend and then an adhock scipy backend which will be faster?(that's the strenght of the flexibility of the backend)

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+    else:
+        return A * B
+
+def new(x, O, M, N):
```
yes, cf: https://github.com/edouardoyallon/kymatio/blob/kymatio-dev-frontend/kymatio/scattering2d/core/scattering2d.py#L23 -- I don't know a shared function for tf, np and torch, do you?

eickenberg replied:
yeah I don't doubt you are using it. I just find the API a bit weird. it seems super specific. If you can describe it in words, could you put a docstring?

edouardoyallon replied:
I added a very short docstring! a longer one maybe? Actually.. due to the way we code backend, I think it could be also nice to do a sort of "backend template", to avoid changing "backends" times the doc string.... and then to specify it for a specific backend.

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+                           'or complex, respectively')
+
+    if inplace:
+        return A.mul_(B)
```
this does exist, and actually what you propose is not a legal np code

edouardoyallon replied:
right  i updated it

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+        output = np.real(np.fft.ifft2(input, norm=None))*input.shape[-1]*input.shape[-2]
+    elif direction == 'C2C':
+        if inverse:
+            output = np.fft.ifft2(input, norm=None)*input.shape[-1]*input.shape[-2]
```
This will change thanks to @nshervt update but I'm not willing to mix different PRs

edouardoyallon commented:
Just wanted to acknowledge @tdeboissiere and @Kegnarok - tf backend added!

edouardoyallon commented on kymatio/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,6 @@
+class Scattering_numpy(object):
```
it allows multiple backends

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, input, k):
```
done

edouardoyallon commented:
hi @eickenberg, I addressed all your concerns - I think the structure of the test is far more readable(it helped me while coding) - please let me know what you think. I'm also happy to remove the config file, it is more explicit now. TF backend is here too. It currently doesn't work on jenkins, I'm not sure why, the bug is not clear. Maybe @janden or @dylex could help?(sorry for the spam!)

eickenberg commented on README.md:
```diff
@@ -74,32 +74,28 @@ Linux and macOS are the two officially supported operating systems.
-
+#### Tensorflow
 
 #### Available backends: PyTorch and scikit-cuda
```
tf?

eickenberg commented on examples/2d/compute_speed.py:
```diff
@@ -25,13 +25,39 @@
+# Determine which devices (CPU or GPU) that are supported by the current
+# backend.
+
+if torch.cuda.is_available():
```
this benchmark either needs to be renamed to `compute_speed_torch` or make the torch/tf import optional using try

edouardoyallon replied:
benchmarks must be backend dependant

eickenberg commented on examples/2d/compute_speed.py:
```diff
@@ -25,13 +25,39 @@
 
 ###############################################################################
-# Finally, we import the `Scattering2D` class that computes the scattering
+# Finally, we import the `Scattering2DTorch` class that computes the scattering
```
We need a default to be occupied by `Scattering2D`. 

I think we can give it dynamic inheritance and say `Scattering2D(frontend='numpy')` or whatever is available by default. (Config file would be optional and not necessary). This will make the examples remain usable and `Scattering2D` do the default thing you want to have happen in most cases.

edouardoyallon replied:
sounds good

eickenberg replied:
Note that this is possible in two ways - I am not sure which is better:
1.
```python
class Scattering2D(object):
    def __init__(self, frontend):
        if frontend == 'pytorch':
            from . import Scattering2DTorch
            self.__base__ = Scattering2DTorch
            super(Scattering2D, self).__init__()
    # other stuff
```

or

```python
class Scattering2D(object):
    def __init__(self, frontend):
        if frontend == 'pytorch':
            from . import Scattering2DTorch
            self.__class__ = Scattering2DTorch
            Scattering2DTorch.__init__(self) # maybe
    # other stuff
```

In the first case, you dynamically inherit, in the second case, you dynamically _become_ that class. What do people think? Semantically probably the second is better?

edouardoyallon replied:
fixed

eickenberg commented on examples/2d/compute_speed.py:
```diff
@@ -25,13 +25,39 @@
 # transform.
 
-from kymatio import Scattering2D
+from kymatio import Scattering2DTorch as Scattering2D
```
So I'd like to see `Scattering2D` still existing, using dynamic inheritance.

lostanlen replied:
and a `logging.info` message: "Using X backend". Like in Keras, except that this one would be easy to turn off by raising the `LOGGING_LEVEL` (Keras uses `print` !)

eickenberg commented on examples/2d/compute_speed.py:
```diff
@@ -59,22 +85,13 @@
-scattering = Scattering2D(J, shape=(M, N), L=L)
+
 
 x = torch.randn(batch_size, 3, M, N, dtype=torch.float32)
```
cannot simply have torch in here without renaming file or making sure we have it

eickenberg commented on kymatio/__init__.py:
```diff
@@ -2,18 +2,23 @@
 __all__ = [
             'Scattering1D',
-            'Scattering2D',
+            'Scattering2DTorch',
```
Would set `__all__` after trying to import stuff and then fill it with what is available

eickenberg commented on kymatio/__init__.py:
```diff
@@ -2,18 +2,23 @@
 from .scattering1d.scattering1d import Scattering1D
-from .scattering2d.scattering2d import Scattering2D
+
+from .scattering2d.frontend import Scattering2DTorch
```
Can't assume that we have torch installed

eickenberg commented on kymatio/frontend/tensorflow_frontend.py:
```diff
@@ -0,0 +1,6 @@
+class ScatteringTensorflow(object):
```
Is there nothing to inherit from? Maybe a `tf.keras` object?

lostanlen replied:
`tf.keras.layers.Layer` with `trainable=False` and `dynamic=False`

https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer

eickenberg commented on kymatio/scattering2d/__init__.py:
```diff
@@ -1,4 +1 @@
-
-
-from .scattering2d import Scattering2D
+from .frontend import *
```
Is it necessary to import `*` or can we be explicit? It should be safe, but if it is avoidable, it might as well be avoided. I also don't yet see what this file contains

lostanlen replied:
i agree

eickenberg commented on kymatio/scattering2d/backend/__init__.py:
```diff
@@ -1,71 +1,25 @@
+    from .tensorflow_backend import backend as tensorflow_backend
+    __all__.append('tensorflow_backend')
+except:
+    pass
```
Could maybe raise an error at the end of this if `__all__` is completely empty

lostanlen replied:
does the order matter here? if so, why not put numpy at the end if it's the slowest?

eickenberg commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,165 @@
+# Authors: Edouard Oyallon, Sergey Zagoruyko
```
Didn't somebody else contribute to this?

edouardoyallon replied:
well, I took some inspirations yet I acknowledged the author. I think mentioning him here would be confusing

eickenberg commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,165 @@
+        self.build()
+
+    def build(self):
+        self.np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))
```
why `np_pad`, when we are in `tf` mode?

eickenberg commented on kymatio/scattering2d/backend/torch_skcuda_backend.py:
```diff
@@ -210,14 +133,14 @@ def __init__(self):
-        if not input.is_cuda:
+    def __call__(self, x):
+        if not x.is_cuda:
             raise RuntimeError('Use the torch backend for cpu tensors!')
```
what about numpy backend? Maybe you want to edit to `'Use the torch backend (without skcuda) for cpu tensors!'`

lostanlen commented on README.md:
```diff
@@ -1,4 +1,4 @@
-Kymatio: Wavelet scattering in PyTorch
+Kymatio: Wavelet scattering in Python
```
This is amazing 😭

lostanlen commented on CONTRIBUTORS.md:
```diff
@@ -13,4 +13,6 @@ Edouard Oyallon
 Sergey Zagoruyko
-Jonas1312
\ No newline at end of file
+Jonas1312
```
i would strongly prefer to stick to a first name - last name basis

edouardoyallon replied:
some people might not want to reveal their name..(i'm fine with it)

lostanlen replied:
it's trivial to find his name from his GitHub handle. this argument doesn't hold

eickenberg replied:
@Jonas1312 Do you have a preference? If not, it would probably be nice to change to first-last for consistency

lostanlen replied:
it's been a month since this. let's make the replacement to first name - last name

edouardoyallon replied:
OK can you do it?

lostanlen commented on CONTRIBUTORS.md:
```diff
@@ -13,4 +13,6 @@ Edouard Oyallon
\ No newline at end of file
+Jonas1312
+Florentin Guth
+Thibault de Boissiere
```
@tdeboissiere wouldn't you prefer "de Boissière" ?

lostanlen commented on README.md:
```diff
@@ -52,7 +52,7 @@ Andreux M., Angles T., Exarchakis G., Leonarduzzi R., Rochette G., Thiry L., Zar
 
 * Python (>= 3.5)
-* PyTorch (>= 0.4)
+* PyTorch (>= 1.1)
```
very true

lostanlen commented on README.md:
```diff
@@ -74,32 +74,28 @@ Linux and macOS are the two officially supported operating systems.
 
 ### GPU acceleration
 
+#### Pytorch
```
PyTorch (note the capital T)

lostanlen commented on README.md:
````diff
@@ -74,32 +74,28 @@ Linux and macOS are the two officially supported operating systems.
-os.environ["KYMATIO_BACKEND"] = "skcuda"
-```
-
+#### Tensorflow
````
TensorFlow (note the capital F)

lostanlen commented on README.md:
```diff
@@ -74,32 +74,28 @@ Linux and macOS are the two officially supported operating systems.
 The user may control the choice of backend at runtime by setting the environment variable `KYMATIO_BACKEND`, or by editing the Kymatio configuration file (`~/.config/kymatio/kymatio.cfg` on Linux).
 
-The two available backends are PyTorch (`torch`) and scikit-cuda (`skcuda`).
+The available backends are PyTorch (`torch`), PyTorch+scikit-cuda (`skcuda`), TensorFlow (`tensorflow`) and Numpy (`numpy`).
```
Oxford comma before "and Numpy"
Numpy is capitalized "NumPy"

lostanlen commented on README.md:
```diff
@@ -74,32 +74,28 @@ Linux and macOS are the two officially supported operating systems.
+The available backends are PyTorch (`torch`), PyTorch+scikit-cuda (`skcuda`), TensorFlow (`tensorflow`) and Numpy (`numpy`).
 
-PyTorch is the default backend in 1D, 2D, and 3D scattering. For applications of the 2D scattering transform to large images (e.g. ImageNet, of size 224x224), however, we recommend the scikit-cuda backend, which is substantially faster than PyTorch.
+Numpy is the default backend in 1D, 2D, and 3D scattering. For applications of the 2D scattering transform to large images (e.g. ImageNet, of size 224x224), however, we recommend the scikit-cuda backend, which is substantially faster than PyTorch.
```
NumPy

lostanlen replied:
substantially faster than NumPy

lostanlen commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+        x: input complex tensor.
+        Returns
+        -------
+        output: a tensor with imaginary part set to 0, real part set equal to
```
this is misleading. The `dtype` of `output` is not complex. There is no imaginary part.

lostanlen commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,176 @@
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then the transform
```
then `inverse` should be `None` by default in the prototype. We shouldn't have a documented default value for a kwarg that we end up overriding in another kwarg.

eickenberg replied:
Seconded - there should be an error instead of changing something automatically

edouardoyallon replied:
I do not understand what you mean, can you provide me an example?

eickenberg replied:
Basically, what this docstring says is that _if `direction='C2R'`, I will ignore what you say in `inverse` and just set it to `True`._ What it should do instead is raise an error when `direction='C2R'` and `inverse=False`.
If you want to keep this somewhat sneaky behavior, then you can do what Vincent suggested and set `inverse=None` as default, which will be interpreted as `False` when `direction='C2C'`, and `True` when `direction='C2R'`.
I think I'd honestly just prefer an error to be raised and no `None` logic to be there

eickenberg replied:
As far as I can tell from the code, this is implemented exactly like that. So all that needs to change is the docstring, saying that `C2R` requires `inverse=True`

edouardoyallon replied:
OH ok.

lostanlen commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,165 @@
+        x: input complex tensor.
+        Returns
+        -------
+        output: a tensor with imaginary part set to 0, real part set equal to
```
same comments as numpy backend

lostanlen commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,165 @@
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then the transform
```
same comment as numpy backend

lostanlen commented on kymatio/scattering2d/backend/torch_backend.py:
```diff
@@ -109,16 +102,17 @@ class SubsampleFourier(object):
+        batch_shape = x.shape[:-3]
+        signal_shape = x.shape[-3:]
+        x = x.view((-1,) + signal_shape)
+        y = x.view(-1,
```
you're a genius ⛄️

janden commented:
So looks like 6a19c1540174df556366a0bdbb74fdc735e4e771 broke the 1D tests, but only in conda? I'm guessing this has something to do with TF being installed there, but potentially not in the pip tests?

lostanlen commented on examples/2d/plot_filters.py:
```diff
@@ -49,7 +49,7 @@ def colorize(z):
 plt.rc('font', family='serif')
 i = 0
 for filter in filters_set['psi']:
```
wait, this thing works? isn't filter a reserved keyword in Python?

lostanlen commented on examples/2d/regularized_inverse_scattering_MNIST.py:
```diff
@@ -94,7 +94,7 @@ def main():
     fixed_batch = fixed_batch[0].float().cuda()
 
-    scattering = Scattering(J=2, shape=(28, 28))
+    scattering = Scattering(J=2, shape=(28, 28), frontend='Torch')
```
lowercase?

eickenberg commented on Jenkinsfile:
```diff
@@ -17,11 +17,12 @@ pipeline {
       }
       steps {
 	sh 'python3 -c "import torch; torch.cuda.current_device()"'
+	sh 'python3 -c "import tensorflow as tf; tf.test.is_gpu_available()"'
```
Can I get a brief explanation for my own education: Is this line here to get a diagnostic, an error message if gpu is not available, but the tests are still attempted, or does failure at this stage entail the script being stopped? (Essentially I am asking the equivalent of "Is `set -e` to be assumed or not")

janden replied:
It was mostly there to force a failure if TF wasn't working properly. Otherwise we'd just get segfaults in the middle of the unit tests without much explanation. It's not strictly necessary but acts as a useful canary.

eickenberg commented on Jenkinsfile:
```diff
@@ -17,11 +17,12 @@ pipeline {
 	  source $HOME/bin/activate
 	  python3 setup.py develop
-	  KYMATIO_BACKEND=$STAGE_NAME python3 -m pytest --cov=kymatio
+	  KYMATIO_BACKEND=$STAGE_NAME CUDA_VISIBLE_DEVICES=0 TF_FORCE_GPU_ALLOW_GROWTH=true python3 -m pytest --cov=kymatio
```
but so this only checks coverage for non-gpu code?

janden replied:
Actually, `CUDA_VISIBLE_DEVICES` forces us to use the first GPU for the tests. The Jenkins node has two GPUs, so to prevent any shenanigans (where it allocates something on one GPU and then tries to use it on the other one), we force it to only use one.

On the other hand, it might be better to leave it unrestricted. May miss some subtle bugs by doing this restriction. At any rate, once we have multi-GPU support, we'll want to use both.

edouardoyallon replied:
thinking of it, I think it could be nice to remove and not to force anything, now that tf works properly

edouardoyallon commented:
> In terms of API, while the current `Scattering2D(..., frontend='blahblah')` is useful, it would be more natural to also include this specification at the import stage, with something like `from kymatio.torch import Scattering2D`. This should be fairly straightforward. It is also a bit strange that the frontend can be specified as a string, while the backend is an object. It would be good if the backend also could be given as a string.

Ok. So in other words, both frontend and backend should be accessible from the `scattering` object via a text argument or via an appropriate import. This sounds reasonable, simple to implement and a good idea.

> Regarding the class structure, there is quite a lot of code that is duplicated between the different frontends (in terms of setting up, various meta tasks, and so on), not to mention the documentation (which is basically the same). While this may work now, I'm worried it will make things hard to maintain, as every change now needs to be duplicated in three locations. Ideally, we would factor out these common parts into smaller helper function, or have all classes inherit from some common `Scattering1D` class (I would lean towards the latter, but am unsure as to how multiple inheritance works in Python, given that we _need_ the inheritance relationships from `torch.nn.Module` and `tf.Module`).

I agree with this statement, in particular for meta functions of the 1D code, such that `loginfo` ( https://github.com/kymatio/kymatio/blob/kymatio-dev-frontend/kymatio/scattering1d/frontend/numpy_frontend.py#L348 ), `outputsize` ( https://github.com/kymatio/kymatio/blob/kymatio-dev-frontend/kymatio/scattering1d/frontend/numpy_frontend.py#L255 ) or `meta` ( https://github.com/kymatio/kymatio/blob/kymatio-dev-frontend/kymatio/scattering1d/frontend/numpy_frontend.py#L241 ). In 2D there is currently no such issues because I did not implement those important primitives. Here is my 2cents view and attempt: I tried to regroup common codes to frontend into the `utils.py` file. I think, indeed, that those routines should be inherited via polymorphism by a 1Dfrontend or 2Dfrontend class. To be more precise, each functional(i.e., working in practice) frontend class should inherit both from its _framework_ (e.g., `torch` ) and its _specifity_ (e.g., `2D`). I think one should use `abc` python polymorphism, e.g., metaclass and abstract implementation. While this can sound like an overkill, this will ensure at 100% that nobody will forget to implement a routine before adding a new scattering.(it's very likely that in 5y, ppl will still create new scattering class) One issue I observed is that currently, the scattering object changes dynamically its class ( https://github.com/kymatio/kymatio/blob/kymatio-dev-frontend/kymatio/scattering1d/frontend/__init__.py#L15 ) and this kills this mechanism. We should also discuss the documentation of inherited class: imho it should be minimum.

> Finally, while there are extensive tests for the Torch frontend, the same tests are missing from the NumPy and TF frontends. As a result, we don't have the code coverage we need there. One solution would be to manually replicate the tests for these frontends, but here we run into a code duplication issue again. It will be a bit more work, but I don't see why the tests couldn't be parametrized over the frontends in addition to the backend and devices. What would be needed is a set of helper functions for creating dummy data and verifying that it's close to some reference.

I agree. Contrary to what I did, the test shouldn't be related to a `frontend` file, but to the `core.py` file and the way the scattering is computed.

MuawizChaudhary commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+    """
+
+    N = x.shape[-1]
+    res = x.reshape(x.shape[:-2] + (k, N // k)).mean(axis=-2, keepdims=True)
```
Changing to `x.shape[:-1]` appears to fail the numpy test. Output: 
`E   ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 4 dimension(s) and the array at index 1 has 5 dimension(s)

<__array_function__ internals>:6: ValueError`

Seems like the shapes are wrong, but I don't think this is adequate enough explanation for why in numpy its `x.shape[:-2]` while in pytorch it's `x.shape[:-1]`.

MuawizChaudhary replied:
I'll try and update tomorrow.

edouardoyallon replied:
This part of the code was well coded: in torch, there is an extra-dimension of lenght 2 that corresponds to tensors... not in numpy

MuawizChaudhary commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+
+    Parameters
+    ----------
+    x : tensor
```
I put down `numpy array` while I was working on the documentation for the numpy backends. Is it more appropriate to have the datatype be `complex128`?

MuawizChaudhary commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+    pad_right : int
+        amount to add on the right of the tensor (at the end of the temporal
+        axis).
+    to_complex : boolean, optional
```
I mention in the docstring on the numpy back end that it does nothing. Do we want to change the behavior of the function on the numpy backend?

edouardoyallon replied:
I think it's fine to keep it as it is.

janden replied:
We can't have documentation that doesn't correspond to what the function does, even if it's mentioned somewhere else. Why not fix this?

MuawizChaudhary commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+    """
+    return np.fft.ifft(x)
+
+def finalize(s0, s1, s2):
```
I think `finalize` seems to make sense... we're 'finalizing' the scattering transform by stacking the coefficients together. Thoughts @edouardoyallon ?

edouardoyallon replied:
I'm pros finalize too

janden replied:
Because, again, it could mean *anything*. It could be applying the lowpass filter to the wavelet modulus coefficients. It could be computing the log (which is a common postprocessing step in 1D scattering for audio). Why use a generic term when a specific one will do?

MuawizChaudhary commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+        Q = 8
+
+        # Generate a sample signal.
+        x = torch.randn(1, 1, T)
```
changed to `np.random.randn(1, 1, T).view(np.complex128)`, but let me check to make sure this works.

MuawizChaudhary commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+        if not self.backend:
+            from ..backend.numpy_backend import backend
+            self.backend = backend
+        elif self.backend.name[0:5] != 'numpy':
```
Reminder to self to do this across all frontends for 123ds.

MuawizChaudhary replied:
I did this and it seems to break all the tests...

MuawizChaudhary commented on kymatio/scattering2d/frontend/tensorflow_frontend.py:
```diff
@@ -0,0 +1,119 @@
+        self.phi, self.psi = filters['phi'], filters['psi']
+
+    def scattering(self, input):
+        with tf.name_scope('scattering') as scope:
```
I believe in the tensorboard computation graph/results 'scattering/' is prefixed to all operation names. See [this](https://www.tensorflow.org/api_docs/python/tf/name_scope) for an even more indepth explanation.

MuawizChaudhary commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+        A tensor of the same size as x containing its Fourier transform in the
+        standard PyTorch FFT ordering.
+    """
+    return np.fft.fft(x)
```
I agree that it seems to be asking for trouble, but I question if the effort to refactor the numpy and tf backends are worth it if we consider that the PyTorch developers are in the process of creating a complex datatype as seen over [here](https://github.com/pytorch/pytorch/issues/755).

The discussion sounds like the datatype may not be added for a while, but when the complex datatype is implemented I question if it would be worth the cost to refactor the codebase to support the complex datatype across all backends. They claim over [here](https://github.com/pytorch/pytorch/issues/1033) that there isn't much community need for complex tensors. I believe that it would be beneficial to add our voice to the conversation, to let them know that there is a need. 

From my point of view there are two ways we could proceed: 

We could refactor the tf and np backends similar to how the torch backend does it, complex tensors/numpy arrays as float64 with a final dim of 2, and correct the documentation. (I would be happy to do this, it would not be bad). When torch releases with the complex datatype, there could be a discussion on if it makes sense to integrate the datatype into the torch backend. My guess is that it would make sense, at which point we would want to refactor the tf and np backends. This seems like more overall amount of work, but we don't know when torch releases its implementation of the complex tensor. So the design/implementation would be the same across all backends for a while. 

Alternatively, we could keep the backends as is, and mention somewhere in the documentation on kymat.io the major differences between the torch backend and the tf/np backends. When the complex tensor datatype release for Torch, we have only one backend to change across *d. This seems like less overall amount of work, but we would have two different sets of designs until the release.

Thoughts @edouardoyallon , @janden , @eickenberg ?

edouardoyallon replied:
> complex128, whereas torch outputs an array with shape x.shape + (2,) and datatype float64

I think a strength of the backend is to be insensitive to this. Basically, the way a type is implemented is a low-level consideration and I think it should remain as it is. As mentionned by @MuawizChaudhary , complex tensor might arrive in torch that would allow us to completely get rid of this extra 2 dimension in torch.

edouardoyallon replied:
Thus I'm pros keeping it as it is!

janden replied:
Ok we can do that. This means that we have to be extra careful when working with different backends, however. More annoying is the fact that the docs for the different backends will no longer be identical (we can't just copy and paste them between one another) so we'll have to maintain them separately.

MuawizChaudhary commented on kymatio/scattering1d/tests/test_tensorflow_scattering1d.py:
```diff
@@ -0,0 +1,62 @@
+        Sx = sess.run(S_tf, feed_dict={x_tf: x})
+        sess.close()
+
+        Warning('Tolerance has been slightly lowered here...')
```
The tolerance is 1e-08 on the torch and np allclose. For some reason the tolerance on tensorflow is 10e-07. @edouardoyallon  Explanation?

edouardoyallon replied:
because it doesn't work, roundoff errors for diff frameworks

edouardoyallon commented on examples/2d/compute_speed_torch.py:
```diff
@@ -25,10 +25,36 @@
-import kymatio.scattering2d.backend as backend
+backends = []
+
+try:
```
Well, it's the same mechanism as to test the backend.

janden replied:
Yes and it's ugly there as well.

edouardoyallon commented on kymatio/scattering2d/tests/test_torch_scattering2d.py:
```diff
@@ -0,0 +1,360 @@
+        for test_shape in test_shapes:
+            x = torch.zeros(test_shape)
+
+            if backend.name == 'torch_skcuda':
```
in order to test lines which are ignored otherwise because some functionality are not implemented.

edouardoyallon commented on kymatio/scattering2d/tests/test_torch_scattering2d.py:
```diff
@@ -0,0 +1,360 @@
+    def test_scattering2d_errors(self, backend, device):
+        S = Scattering2D(3, (32, 32), backend=backend, frontend='torch')
+
+        if backend.name == 'torch_skcuda':
```
in order to test lines which are ignored otherwise because some functionality are not implemented.

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,171 @@
+        if self.pre_pad:
+            return x
+        else:
+            np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))
```
It will cause some problems but I think it's safer to handle it later, that's too much work for now. Let's split this task.

edouardoyallon commented:
@janden it seems that @MuawizChaudhary took care of everything you required in your review.

edouardoyallon commented on kymatio/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,25 @@
+       """ This function should call the functional scattering."""
+
+
+   def __call__(self, x):
```
call is typically native for np modules
forward is native for torch module
call is native in tf

edouardoyallon commented on kymatio/frontend/torch_frontend.py:
```diff
@@ -0,0 +1,25 @@
+       self.filters = self.create_and_register_filters() """
+
+   @abstractmethod
+   def create_and_register_filters(self):
```
because one day it will be simply named "create".. for now, i think it's easier to keep them separate because the regiter is a "hack"

janden replied:
So let's just call it `create_filters`, then.

edouardoyallon commented on Jenkinsfile:
```diff
@@ -17,11 +17,12 @@ pipeline {
 	  source $HOME/bin/activate
 	  python3 setup.py develop
-	  KYMATIO_BACKEND=$STAGE_NAME python3 -m pytest --cov=kymatio
+	  KYMATIO_BACKEND=$STAGE_NAME CUDA_VISIBLE_DEVICES=0,1 TF_FORCE_GPU_ALLOW_GROWTH=true python3 -m pytest --cov=kymatio
```
I followed your suggestion

edouardoyallon commented on examples/2d/compute_speed_torch.py:
```diff
@@ -90,40 +107,41 @@
-    fmt_str = '==> Testing Float32 with {} backend, on {}, forward'
-    print(fmt_str.format(backend.NAME, device.upper()))
+for backend in backends:
+    scattering = Scattering2D(J, shape=(M, N), L=L, backend=backend, frontend='torch')
```
fixed

edouardoyallon commented on examples/2d/compute_speed_torch.py:
```diff
@@ -90,40 +107,41 @@
-    else:
-        scattering.cpu()
-        x = x.cpu()
+        if not(device == 'cpu' and backend.name == 'torch_skcuda'):
```
Well, it's better than before I think. We could, but ultimately, skcuda should be compatible with any devices i think.

janden replied:
Are you saying that the skcuda backend should fall back on CPU torch when called on CPU tensors?

edouardoyallon commented on examples/2d/plot_filters.py:
```diff
@@ -48,8 +48,8 @@ def colorize(z):
 i = 0
-for filter in filters_set['psi']:
-    f = filter[0][..., 0]
+for fil in filters_set['psi']:
```
it's a keyword of python

janden replied:
It's not a keyword, it's a builtin function. There's no reason you can't use it as a name (and we did). If you don't want to use `filter`, we can use `f` (and change `f` below to `f0`).

edouardoyallon commented on examples/2d/plot_scattering_disk.py:
```diff
@@ -45,16 +43,12 @@
 src_img = src_img.astype(np.float32) / 255.
-src_img_tensor = torch.from_numpy(src_img).to(device)
-scattering_coefficients = scattering(src_img_tensor)
+scattering_coefficients = scattering(src_img)
```
fixed

edouardoyallon commented on kymatio/__init__.py:
```diff
@@ -1,7 +1,8 @@
 ### Snippet copied from sklearn.__init__
-import warnings
 import re
+import warnings
```
mistake but that's ok leave, no?

edouardoyallon commented on kymatio/backend/fake_backend.py:
```diff
@@ -0,0 +1,5 @@
+from collections import namedtuple
```
well, it's a global backend shared accross 1-2-3D...

edouardoyallon commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+    pad_right : int
+        amount to add on the right of the tensor (at the end of the temporal
+        axis).
+    mode : string, optional
```
WDYM? not used?

janden replied:
Yes.

edouardoyallon commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+    mode : string, optional
+        Padding mode. Options include 'constant' and 'reflect'. See the
+        PyTorch API for other options.  Defaults to 'constant'.
+    value : float, optional
```
WDYM? not used?

janden replied:
Yes.

edouardoyallon commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,213 @@
+    return np.concatenate([np.concatenate(s0, axis=-2), np.concatenate(s1, axis=-2), np.concatenate(s2, axis=-2)], axis=-2)
+
+
+backend = namedtuple('backend', ['name', 'modulus_complex', 'subsample_fourier', 'real', 'unpad', 'fft1d_c2c', 'ifft1d_c2c'])
```
color and taste, imho

janden replied:
I don't understand. Why do something non-standard unless we have a good reason? Inheritance is the standard way of grouping classes with common functionality in OOP.

edouardoyallon commented on kymatio/scattering1d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,216 @@
+    pad_right : int
+        amount to add on the right of the tensor (at the end of the temporal
+        axis).
+    mode : string, optional
```
WDYM? not used?

janden replied:
Yes. Changing the value doesn't change what the function does.

edouardoyallon commented on kymatio/scattering1d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,216 @@
+    mode : string, optional
+        Padding mode. Options include 'constant' and 'reflect'. See the
+        PyTorch API for other options.  Defaults to 'constant'.
+    value : float, optional
```
WDYM? not used?

janden replied:
Yes.

edouardoyallon commented on kymatio/scattering1d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,216 @@
+    """
+    return tf.signal.ifft(x)
+
+def finalize(s0, s1, s2):
```
disagree

edouardoyallon commented on kymatio/scattering1d/backend/torch_backend.py:
```diff
@@ -305,3 +306,24 @@ def ifft1d_c2c(x):
+    if len(s2)>0:
+        return torch.cat([torch.cat(s0, -2), torch.cat(s1, -2), torch.cat(s2, -2)], -2)
+    else:
+        return torch.cat([torch.cat(s0, -2), torch.cat(s1, -2)], -2)
```
it should, this was a bug.

edouardoyallon commented on kymatio/scattering1d/core/scattering1d.py:
```diff
@@ -0,0 +1,159 @@
+        kJ = max(J - oversampling, 0)
+        temporal_size = ind_end[kJ] - ind_start[kJ]
+        #S = x.new(batch_size, sum(size_scattering), temporal_size).fill_(0.)
+        out_S_0, out_S_1, out_S_2 = [], [], []
```
it's to make the software compatibel with everybackends. I couldn't think of a better solution and would behappy if you propose a simpler PR.

edouardoyallon commented on kymatio/scattering1d/frontend/__init__.py:
```diff
@@ -0,0 +1,38 @@
+__all__ = ['Scattering1D']
+
+class Scattering1D(object):
+    def __init__(self, *args, **kwargs):
```
certainly not, this is just inherited by `nn.Module`  now, the software should not handle low level operations.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+__all__ = ['Scattering1DNumpy']
+
+class Scattering1DNumpy(ScatteringNumpy):
+    """The 1D scattering transform
```
I agree with this statements. bankfilters also should be unified. Everything should be unified. but for now it was the simplest(the amount of time will be significant and it'll be easier to split this in a future PR)

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+    GPU, if available. A `Scattering1D` object may be transferred from one
+    to the other using the `cuda()` and `cpu()` methods.
+
+    Given an input Tensor `x` of size `(B, T)`, where `B` is the number of
```
Clever templating.(later)

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+                 oversampling=0, vectorize=True, backend=None):
+        super(Scattering1DNumpy, self).__init__()
+        # Store the parameters
+        self.J = J
```
I agree.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+        elif self.backend.name[0:5] != 'numpy':
+            raise RuntimeError('This backend is not supported.')
+
+        self.r_psi = math.sqrt(0.5)
```
I agree but it's hard to have the good overview.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+        self.psi2_f = psi2_f
+        self.phi_f = phi_f
+
+    def meta(self):
```
I kind of agree. part of the call is in util now.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+    def meta(self):
+        """Get meta information on the transform
+
+        Calls the static method `compute_meta_scattering()` with the
```
right this should be changed.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+    def output_size(self, detail=False):
+        """Get size of the scattering transform
+
+        Calls the static method `precompute_size_scattering()` with the
```
right.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+        return compute_meta_scattering(
+            self.J, self.Q, max_order=self.max_order)
+
+    def output_size(self, detail=False):
```
right.

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+    def scattering(self, x):
+        """Apply the scattering transform
+
+        Given an input Tensor of size `(B, T0)`, where `B` is the batch
```
right

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+
+        Parameters
+        ----------
+        x : tensor
```
right

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+        Returns
+        -------
+        S : tensor or dictionary
+            If the `vectorize` flag is `True`, the output is a Tensor
```
right

edouardoyallon commented on kymatio/scattering1d/frontend/numpy_frontend.py:
```diff
@@ -0,0 +1,346 @@
+            is `False`, it is a dictionary indexed by tuples of filter indices.
+        """
+        # basic checking, should be improved
+        if len(x.shape) < 1:
```
right

edouardoyallon commented on kymatio/scattering1d/frontend/tensorflow_frontend.py:
```diff
@@ -0,0 +1,351 @@
+    $S_J^{(1)} x$, and $S_J^{(2)} x$ or just $S_J^{(0)} x$
+    and $S_J^{(1)} x$.
+
+    The scattering transform may be computed on the CPU (the default) or a
```
absolutely no clue

edouardoyallon commented on kymatio/scattering1d/tests/test_filters_scattering1d.py:
```diff
@@ -1,16 +1,9 @@
-from kymatio.scattering1d.filter_bank import calibrate_scattering_filters
-from kymatio.scattering1d.filter_bank import get_max_dyadic_subsampling
-from kymatio.scattering1d.filter_bank import gauss_1d
+from kymatio.scattering1d.filter_bank import adaptive_choice_P, periodize_filter_fourier, get_normalizing_factor, \
```
13 line > 1 line

edouardoyallon commented on kymatio/scattering1d/tests/test_numpy_scattering1d.py:
```diff
@@ -0,0 +1,53 @@
+import os
```
a bit difficult to add all of them, maybe in a future PR?

edouardoyallon commented on kymatio/scattering1d/tests/test_numpy_scattering1d.py:
```diff
@@ -0,0 +1,53 @@
+from kymatio.scattering1d import Scattering1D
+import torch
+
+class TestScattering1DNumpy:
```
simpler

edouardoyallon commented on kymatio/scattering1d/tests/test_numpy_scattering1d.py:
```diff
@@ -0,0 +1,53 @@
+    def test_Scattering2D(self):
+        test_data_dir = os.path.dirname(__file__)
+        test_data_filename = os.path.join(test_data_dir, 'test_data_1d.pt')
+        data = torch.load(test_data_filename, map_location='cpu')
```
future PR

edouardoyallon commented on kymatio/scattering1d/tests/test_tensorflow_scattering1d.py:
```diff
@@ -0,0 +1,62 @@
+import torch
+import tensorflow as tf
+
+class TestScattering1DTensorflow:
```
same as above

edouardoyallon commented on kymatio/scattering1d/tests/test_tensorflow_scattering1d.py:
```diff
@@ -0,0 +1,62 @@
+    def test_Scattering1D(self):
+        test_data_dir = os.path.dirname(__file__)
+        test_data_filename = os.path.join(test_data_dir, 'test_data_1d.pt')
+        data = torch.load(test_data_filename, map_location='cpu')
```
same as above

edouardoyallon commented on kymatio/scattering1d/utils.py:
```diff
@@ -108,3 +64,216 @@ def compute_padding(J_pad, T):
+        A dictionary with the following keys:
+
+        - `'order`' : tensor
+            A Tensor of length `C`, the total number of scattering
```
right

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,171 @@
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        out = np.zeros((x.shape[0], x.shape[1], x.shape[2] // k, x.shape[3] // k), dtype=x.dtype)
```
future PR

edouardoyallon commented on kymatio/scattering2d/backend/numpy_backend.py:
```diff
@@ -0,0 +1,171 @@
+
+
+
+def fft(x, direction='C2C', inverse=False):
```
strong disagree

edouardoyallon commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,159 @@
+            return x
+        else:
+            paddings = [[0, 0]]*len(x.shape[:-2].as_list())
+            paddings += [[self.pad_size[0],self.pad_size[1]],[self.pad_size[2],self.pad_size[3]]]
```
future pr

edouardoyallon commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,159 @@
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        y = tf.reshape(x,(-1,x.shape[1],
```
future pr

edouardoyallon commented on kymatio/scattering2d/backend/tensorflow_backend.py:
```diff
@@ -0,0 +1,159 @@
+        C : tensor
+            output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
+    """
```
no idea how tf works

edouardoyallon commented:
@MuawizChaudhary I didn't solve some of the issue found by @janden but feel free if you have some times(solved comments do not need to be addressed anymore)

thanks!!
